src/terminal/adapter/adaptDispatch.cpp

```diff
@@ -3656,6 +3656,353 @@ void AdaptDispatch::_ReportDECSACESetting() const
+             << renditionAttributes << ';'
+             << characterAttributes << ';'
+             << flags << ';'
+             << leftSetNumber << ';'
+             << rightSetNumber << ';'
+             << charsetSizes << ';'
+             << charset0 << charset1 << charset2 << charset3
+             << L"\033\\"sv; // An ST ends the sequence.
+
+    const auto response_string = response.str();
+    _api.ReturnResponse({ response_string.data(), response_string.size() });
+}
+
+// Method Description:
+// - DECCIR - This is a parser for the Cursor Information Report received via DECRSPS.
+// Arguments:
+// - <none>
+// Return Value:
+// - a function to parse the report data.
+ITermDispatch::StringHandler AdaptDispatch::_RestoreCursorInformation()
+{
+    // clang-format off
+    enum Field { Row, Column, Page, SGR, Attr, Flags, GL, GR, Sizes, G0, G1, G2, G3 };
+    // clang-format on
+    constexpr til::enumset<Field> numeric{ Field::Row, Field::Column, Field::Page, Field::GL, Field::GR };
+    constexpr til::enumset<Field> flags{ Field::SGR, Field::Attr, Field::Flags, Field::Sizes };
+    constexpr til::enumset<Field> charset{ Field::G0, Field::G1, Field::G2, Field::G3 };
+    struct State
+    {
+        Field field{ Field::Row };
+        VTInt value{ 0 };
+        VTIDBuilder charsetId{};
+        std::array<bool, 4> charset96{};
+        VTParameter row{};
+        VTParameter column{};
+    };
+    auto& textBuffer = _api.GetTextBuffer();
+    return [&, state = State{}](const auto ch) mutable {
+        if (numeric.test(state.field))
+        {
+            if (ch >= '0' && ch <= '9')
+            {
+                state.value *= 10;
+                state.value += (ch - L'0');
+                state.value = std::min(state.value, MAX_PARAMETER_VALUE);
+            }
+            else if (ch == L';' || ch == AsciiChars::ESC)
+            {
+                if (state.field == Field::Row)
+                {
+                    state.row = state.value;
+                }
+                else if (state.field == Field::Column)
+                {
+                    state.column = state.value;
+                }
+                else if (state.field == Field::Page)
+                {
+                    // Paging is not supported yet (GH#13892).
+                }
+                else if (state.field == Field::GL && state.value <= 3)
+                {
+                    LockingShift(state.value);
+                }
+                else if (state.field == Field::GR && state.value <= 3)
+                {
+                    LockingShiftRight(state.value);
+                }
+                state.value = {};
+                state.field = static_cast<Field>(state.field + 1);
+            }
+        }
+        else if (flags.test(state.field))
+        {
+            if (ch >= L'@' && ch <= '~' && !state.value)
+            {
+                state.value = ch;
+                if (state.field == Field::SGR)
```

Comment:
nit: I think it would make sense to move this if/else chain down to the `L';'` branch, because then it would be similar to other parsing techniques ("accumulate state, then flush state when encountering a terminator") and match the `numeric` branch above.
Also, I wonder if using a `switch` would make sense here... But these are just some nits, while marveling at this parser.

Reply:
I'm honestly not sure about this particular operation, but from past experience with other `DCS` sequences, it's generally expected that the content is processed as soon as it's received. So if the sequence is aborted before we get a `;` this field would still be processed.

As for the `switch`, I started like that, but in some of the cases above I needed an additional condition, and using an `if` just made that a whole lot neater. That wasn't necessary here, but I wanted to keep the code consistent.

Reply:
Yeah, sorry. I definitely should have commented something here.